File: src/eks_04_List.cpp
/**
 *   Programeksempel nr 4 - List (selvlaget enkel klasse).
 *
 *   Eksemplet viser en selvlaget implementasjon av container-klassen List.
 *   Det er laget kode for følgende funksjoner:
 *
 *     -  List()
 *     -  ~List
 *     -  void  display()
 *     -  bool  empty()
 *     -  bool  find(const T t)
 *     -  void  insert(const T t)
 *     -  bool  remove(const T t)
 *     -  T     removeNo(const int no)   // Fra 1-en (ikke 0-null) og oppover!
 *     -  int   size()
 *
 *   NB: For å formodentlig øke leseligheten og oversikten er medlems-
 *       funksjonene bevisst IKKE kommentert etter Doxygen-standarden.
 *
 *   @file     EKS_04_List.CPP
 *   @author   Frode Haug, NTNU
 */


#include <iostream>                  //  cout
#include <string>                    //  string
using namespace std;


/**
 *  Container-klassen List, som er en selvlaget versjon av STLs <list>.
 *
 *  Inneholder en privat struct 'Node', kun med ett datamedlem og en peker til
 *  neste i listen. I tillegg er det en peker ('hode') til starten på lista,
 *  og antall noder i listen. Lista er sortert på 'T' i stigende rekkefølge.
 */
template <typename T>                //  'T': fleksibel datatype!
class List {

  private:
      // NB  NB  NB:  "Java-struct" - se HELT TIL SLUTT på denne filen!
      struct Node {                  //  Skjult struct for liste med 'data'.
          T  data;                   //  Unnalagrede 'data'.
          Node* neste;               //  Peker til neste i listen.
          Node(const T d) { data = d;  neste = nullptr; }  //  Constructor.
      };

      Node*  hode;                 //  Dummy hode for å forenkle mye av koden.
      int    antallNoder;          //  Nåværende antall i listen.

  public:                            //  Initierer medlemmene:
      List(const T t)  {  hode = new Node(t);  antallNoder = 0;  }

      ~List()  {                     //  Destructor - sletter HELE listen:
          Node* np;                  //  Hjelpepeker til neste i listen.
          while (hode) {             //  Så lenge noe å slette:
              np = hode->neste;      //  Peker til den andre noden.
              delete hode;           //  Sletter den første noden.
              hode = np;             //  Setter 'hode' til å peke på
          }                          //     den nye første noden.
      }


      void display() const  {        //  Skriver listens innhold:
          Node* np = hode->neste;    //  Hopper over selve hodet.
          while (np) {               //  Ikke nådd listens slutt:
              cout << '\t' << np->data << '\n';  //  Skriver nodens data.
              np = np->neste;        //  Hopper/går til neste i listen.
          }
          cout << "\n\tAntall noder i listen:  " << antallNoder << '\n';
      }

                                     //  Returnerer om listen er tom eller ei:
      bool empty() const  {  return (antallNoder == 0);  }


      bool find(const T t) const  {  // Returnerer om 't' er å finne eller ei:
          Node* np = hode->neste;       //  Hopper over selve hodet.
          while (np  &&  np->data < t)  //  Blar til der noe evt. ligger:
              np = np->neste;
          return (np  &&  np->data == t);  //  Funnet eller ei?
      }


      void insert(const T t)  {      //  Setter inn ny node sortert:
          Node* np = hode,           //  Initierer til selve hodet.
              * nyNode = new Node(t);  //  Oppretter den nye noden.

                            //  Neste finnes, og den er < enn den nyeste
                            //    (dvs. ved identiske legges den nye forrest):
           while (np->neste  &&  np->neste->data < nyNode->data)
                  np = np->neste;      //  Blar frem til neste node.

           nyNode->neste = np->neste;  //  Hekter inn mellom to noder
           np->neste = nyNode;         //    (evt. helt til slutt):
           antallNoder++;              //  Øker totalantallet noder.
      }


      bool remove(const T t) {         //  Returnerer om fjernet eller ei:
          Node* np = hode,             //  Initierer til selve hodet.
              * np2 = hode->neste;     //  Initierer til etter hodet.
          while (np2  &&  np2->data < t) {  //  Blar til der evt. ligger:
              np = np2;               //  Hopper hakket etter.
              np2 = np2->neste;       //  Blar til neste.
          }

          if (np2  &&  np2->data == t) {  //  Funnet:
              np->neste = np2->neste;     //  Hekter ut 'np2':
              delete np2;                 //  Sletter noden 'np2'.
              antallNoder--;              //  Minsker totalantallet noder.
              return true;                //  Retunerer at er fjernet.
          }
          else                            //  Retunerer at er IKKE fjernet.
              return false;
      }


      T removeNo(const int no) {       //  Returnerer (om mulig) node nr.'no':
          Node* np = hode,                //  Initierer til selve hoet.
              * np2 = hode->neste;        //  Initierer til etter hodet.
          T returData;                    //  Evt. funnet nodes 'data'.
          if (no >= 1 && no <= antallNoder) {  //  'no' i rett intervall:
              for (int i = 1; i < no; i++) {  //  Blar frem til nr.'no':
                  np = np2;    np2 = np2->neste;  // Pekerne flyttes fremover:
              }
              np->neste = np2->neste;     //  Hekter ut 'np2':
              returData = np2->data;      //  Tar vare på dens 'data'.
              delete np2;                 //  Sletter noden 'np2'.
              antallNoder--;              //  Minsker totalantallet noder.
              return returData;           //  Returnerer unnalagrede 'data'.
          }
          else
              cout << "\nNumber out of bounds .....\n\n"; // Egen melding ??
          return T();
      }

                                         //  Returnerer antall noder i listen:
      int size() const  {  return antallNoder;  }
};


/**
 *  Hovedprogrammet:
 */
int main()  {
    List <string> liste("");           //  NB:   vs.  list <string> liste;
                                       //        fra OOProg, da med liten 'l'.

    liste.insert("Oslo");   liste.insert("Bergen");     liste.insert("Molde");
    liste.insert("Nas");    liste.insert("Armenistis"); liste.insert("Hamar");
    liste.display();

    cout << "\nHamar finnes";    if (!liste.find("Hamar"))  cout << " IKKE!";
    cout << "\nLarvik finnes";   if (!liste.find("Larvik")) cout << " IKKE!";
    cout << '\n';

    if (!liste.remove("Molde"))  cout << "Å fjerne 'Molde' mislyktes!\n";
    if (!liste.remove("Oslo"))   cout << "Å fjerne 'Oslo' mislyktes!\n";
                                             //  "Gjøvik" finnes ikke:
    if (!liste.remove("Gjøvik")) cout << "Å fjerne 'Gjøvik' mislyktes!\n";

    liste.display();

    cout << "Fjernet nr.3: " << liste.removeNo(3) << '\n';   liste.display();
    cout << "Fjernet nr.3: " << liste.removeNo(3) << '\n';   liste.display();
    cout << "Fjernet nr.1: " << liste.removeNo(1) << '\n';   liste.display();

                                            //  Nr.4 ikke finnes:
    liste.removeNo(4);  cout << "Fjernet nr.4:\n ";   liste.display();


// ======================   LISTE MED INT'er:   ==============================

    List <int> liste2(0);
    liste2.insert(29);     liste2.insert(17);    liste2.insert(22);
    liste2.display();

    if (liste2.remove(312) == 0)  cout << "\n\nFant ikke 312!\n";
    if (liste2.removeNo(6) == 0)  cout << "Fant ikke nr.6!\n";

    cout << "\n\n";
    return 0;
}


/*****************************************************************************

  "Java-struct":

   public class Node  {
       public  int/char/String  data;
       public  Node neste;                        //  Constructor:
       public  Node (const  int/char/String  d)  { data = d;  neste =  null; }
   }

*****************************************************************************/

File: src/eks_04_List.java
/**
 *   Programeksempel nr 4 - List (selvlaget enkel klasse).
 *
 *   Eksemplet viser en selvlaget implementasjon av container-klassen List.
 *   Det er laget kode for følgende funksjoner:
 *
 *     -  List()
 *     -  void     display()
 *     -  boolean  empty()
 *     -  boolean  find(final T t)
 *     -  void     insert(final T t)
 *     -  boolean  remove(final T t)
 *     -  T        removeNo(final int no)   // Fra 1-en (ikke 0-null) og oppover!
 *     -  int      size()
 *
 *
 *   Generelt om forskjellen ift. C++-koden:  Se HELT til slutt i denne filen.
 *
 *   @file     EKS_04_List.java
 *   @author   Marius Lindvall, NTNU
 */


/**
 *  Container-klassen List, som er en selvlaget versjon av STLs <list> i C++.
 *
 *  Inneholder en privat class 'Node', kun med ett datamedlem og en referanse til
 *  neste i listen. I tillegg er det en referanse ('hode') til starten på lista,
 *  og antall noder i listen. Lista er sortert på 'T' i stigende rekkefølge.
 */
public class EKS_04_List {
    static class List<T extends Comparable<T>> {  //  'T': fleksibel datatype!
        private class Node {
            public T data;              //  Unnalagrede 'data'.
            Node neste;                 //  Peker til neste i listen.
            Node(final T d) { data = d;  neste = null; }  //  Constructor
        }

        private Node  hode;             //  Dummy hode for å forenkle mye av koden.
        private int   antallNoder;      //  Nåværende antall i listen.

                                        //  Initierer medlemmene:
        public List(final T t) { hode = new Node(t); antallNoder = 0; }
        
        public void display() {         //  Skriver listens innhold:
            Node np = hode.neste;       //  Hopper over selve hodet.
            while (np != null) {        //  Ikke nådd listens slutt:
                System.out.println("\t" + np.data);  //  Skriver nodens data.
                np = np.neste;          //  Hopper/går til neste i listen.
            }
            System.out.println("\n\tAntall noder i listen:  " + antallNoder);
        }
        
                                        //  Returnerer om listen er tom eller ei:
        public boolean empty() { return antallNoder == 0; }
        
                                        //  Returnerer om 't' er å finne eller ei:
        public boolean find(final T t) {
            Node np = hode.neste;       //  Hopper over selve hodet.
                                        //  Blar til der noe evt. ligger:
            while (np != null && np.data.compareTo(t) < 0)
                np = np.neste;
            return np != null && np.data.equals(t);  //  Funnet eller ei?
        }
        
        public void insert(final T t) { //  Setter inn ny node sortert:
            Node np = hode,             //  Initierer til selve hodet.
                 nyNode = new Node(t);  //  Oppretter den nye noden.
            
                        //  Neste finnes, og den er < enn den nyeste
                        //      (dvs. ved identiske legges den nye forrest):
            while (np.neste != null && np.neste.data.compareTo(nyNode.data) < 0)
                np = np.neste;          //  Blar frem til neste node.
            
            nyNode.neste = np.neste;    //  Hekter inn mellom to noder
            np.neste = nyNode;          //    (evt. helt til slutt):
            antallNoder++;              //  Øker totalantallet noder.
        }
        
        public boolean remove(final T t) {  //  Returnerer om fjernet eller ei:
            Node np = hode,             //  Initierer til selve hodet.
                 np2 = hode.neste;      //  Initierer til etter hodet.
                                        //  Blar til der evt. ligger:
            while (np2 != null && np2.data.compareTo(t) < 0) {
                np = np2;               //  Hopper hakket etter.
                np2 = np2.neste;        //  Blar til neste.
            }
            
            if (np2 != null && np2.data.equals(t)) {  //  Funnet:
                np.neste = np2.neste;   //  Hekter ut 'np2':
                np2.neste = null;
                antallNoder--;          //  Minsker totalantallet noder.
                return true;            //  Retunerer at er fjernet.
            }
            else                        //  Retunerer at er IKKE fjernet.
                return false;
        }
        
        public T removeNo(final int no) {  //  Returnerer (om mulig) node nr.'no':
            Node np = hode,             //  Initierer til selve hodet.
                 np2 = hode.neste;      //  Initierer til etter hodet.
            if (no >= 1 && no <= antallNoder) {  //  'no' i rett intervall:
                for (int i = 1; i < no; i++) {  //  Blar frem til nr.'no': 
                    np = np2;    np2 = np2.neste;  // Pekerne flyttes fremover:
                }
                np.neste = np2.neste;   //  Hekter ut 'np2':
                np2.neste = null;
                antallNoder--;          //  Minsker totalantallet noder.
                return np2.data;        //  Returnerer nodens 'data'.
            } else
                                        //  Egen melding ??
                System.out.println("\nNumber out of bounds .....\n");
            return null;
        }
        
                                        //  Returnerer antall noder i listen:
        public int size() { return antallNoder; }
    }


    
    /**
     * Hovedprogrammet:
     */
    public static void main(String[] args) {
        List<String> liste = new List<>("");
        
        liste.insert("Oslo");   liste.insert("Bergen");     liste.insert("Molde");
        liste.insert("Nas");    liste.insert("Armenistis"); liste.insert("Hamar");
        liste.display();
        
        System.out.print("\nHamar finnes");
        if (!liste.find("Hamar"))  System.out.print(" IKKE!");
        System.out.print("\nLarvik finnes");
        if (!liste.find("Larvik")) System.out.print(" IKKE!");
        System.out.print('\n');
        
        if (liste.remove("Molde") == null)
            System.out.println("Å fjerne 'Molde' mislyktes!");
        if (liste.remove("Oslo")  == null)
            System.out.println("Å fjerne 'Oslo' mislyktes!");
        if (liste.remove("Gjøvik") == null)
            System.out.println("Å fjerne 'Gjøvik' mislyktes!");
        
        liste.display();

        System.out.print("Fjernet nr.3: " + liste.removeNo(3));   liste.display();
        System.out.print("Fjernet nr.3: " + liste.removeNo(3));   liste.display();
        System.out.print("Fjernet nr.1: " + liste.removeNo(1));   liste.display();
        
        liste.removeNo(4);  System.out.println("Fjernet nr.4: ");   liste.display();
        
// ============================   LISTE MED INT'er:   ==================================
        
        List<Integer> liste2 = new List<>(0);
        liste2.insert(29);  liste2.insert(17);  liste2.insert(22);
        liste2.display();
        
        if (liste2.remove(312) == null)  System.out.println("\n\nFant ikke 312!");
        if (liste2.removeNo(6) == null)  System.out.println("Fant ikke nr.6!");
        
        System.out.println("\n\n");
    }
}


/*************************************************************************************** 

 Noen viktige og påkrevde forskjeller ift. C++-versjonen:

 - Destructor finnes ikke i Java på samme måte som i C++; garbage collector tar seg av 
   rydding i minnet automatisk, så dette er utelatt.
 - Java har også generics som likner på templates i C++, så generics er brukt. "return 0;"
   er ikke gyldig i Java når funksjonen er definert med en annen return type enn tall.
   "return null;" er brukt i stedet, som tilsvarer nullptr.
 - < og > kan ikke brukes for å sammenligne objekter i Java, så der det er brukt generics
   er det satt at `T extends Comparable` og .compareTo brukt i stedet for < og >.

****************************************************************************************/
